cli/wallet/file/file.go

```diff
@@ -0,0 +1,348 @@
+	cfg := make(map[string]interface{})
+	cfg[cfgAlgorithm], _ = wf.flags.GetString(cfgAlgorithm)
+	return cfg, nil
+}
+
+func (wf *fileWalletFactory) RequiresPassphrase() bool {
+	// A file-backed wallet always requires a passphrase.
+	return true
+}
+
+func (wf *fileWalletFactory) unmarshalConfig(raw map[string]interface{}) (*walletConfig, error) {
+	if raw == nil {
+		return nil, fmt.Errorf("missing configuration")
+	}
+
+	var cfg walletConfig
+	if err := mapstructure.Decode(raw, &cfg); err != nil {
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+func (wf *fileWalletFactory) Create(name string, passphrase string, rawCfg map[string]interface{}) (wallet.Wallet, error) {
+	cfg, err := wf.unmarshalConfig(rawCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate entropy.
+	entropy, _ := bip39.NewEntropy(256)
+	mnemonic, _ := bip39.NewMnemonic(entropy)
+	state := &secretState{
+		Algorithm: cfg.Algorithm,
+		Data:      mnemonic,
+	}
+
+	// Seal state.
+	envelope, err := state.Seal(passphrase)
+	if err != nil {
+		return nil, fmt.Errorf("failed to seal state: %w", err)
+	}
+
+	raw, err := json.Marshal(envelope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal envelope: %w", err)
+	}
+	if err := ioutil.WriteFile(getWalletFilename(name), raw, 0o600); err != nil {
+		return nil, fmt.Errorf("failed to save state: %w", err)
+	}
+
+	// Create a proper wallet based on the chosen algorithm.
+	return newWallet(state, cfg)
+}
+
+func (wf *fileWalletFactory) Load(name string, passphrase string, rawCfg map[string]interface{}) (wallet.Wallet, error) {
+	cfg, err := wf.unmarshalConfig(rawCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Load state from encrypted file.
+	raw, err := ioutil.ReadFile(getWalletFilename(name))
+	if err != nil {
+		return nil, fmt.Errorf("failed to load wallet state: %w", err)
+	}
+
+	var envelope secretStateEnvelope
+	if err = json.Unmarshal(raw, &envelope); err != nil {
+		return nil, fmt.Errorf("failed to load wallet state: %w", err)
+	}
+
+	var state *secretState
+	if state, err = envelope.Open(passphrase); err != nil {
+		return nil, fmt.Errorf("failed to open wallet state (maybe incorrect passphrase?)")
+	}
+
+	return newWallet(state, cfg)
+}
+
+func (wf *fileWalletFactory) Remove(name string, rawCfg map[string]interface{}) error {
+	return os.Remove(getWalletFilename(name))
+}
+
+func (wf *fileWalletFactory) Import(name string, passphrase string, rawCfg map[string]interface{}, src *wallet.ImportSource) (wallet.Wallet, error) {
+	cfg, err := wf.unmarshalConfig(rawCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	var state secretState
+	switch src.Kind {
+	case wallet.ImportKindMnemonic:
+		// Import a mnemonic.
+		state.Algorithm = cfg.Algorithm
+		state.Data = src.Data
+	case wallet.ImportKindPrivateKey:
+		// Import a private key. This requires us to change the algorithm to the raw variant.
+		switch cfg.Algorithm {
+		case algorithmEd25519:
+			state.Algorithm = algorithmEd25519Raw
```

Comment:
we use a special value here instead of, say storing the ImportSource.kind alongside. what does this get us? maybe in the future we can change the default derivation scheme used when the user only specifies "mnemonic"?